bucket.go

```diff
@@ -35,6 +35,22 @@ func (b *bucket[T]) get(key string) *Item[T] {
 	return b.lookup[key]
 }
 
+func (b *bucket[T]) setnx(key string, value T, duration time.Duration, track bool) *Item[T] {
+	b.Lock()
+	defer b.Unlock()
+
+	item := b.lookup[key]
+	if item != nil {
+		return item
+	}
+
+	expires := time.Now().Add(duration).UnixNano()
```

Comment:
I'm ok with this, but...did I can think of two slight tweaks.

The first is that `expires` and `item` are setup BEFORE the lock. This would still be atomic,  it would reduce the work that's done under the lock, but would be wasteful when the item already exists:

```go
expires := time.Now().Add(duration).UnixNano()
new_item := newItem(key, value, expires, track)

b.Lock()
defer b.Unlock()

item := b.lookup[key]
if item != nil {
  return item
}

b.lookup[key] = new_item
return item
```

The other option is to use a read lock + write lock, requiring to double check the lookup:

```go

b.RLock()
item := b.lookup[key]
b.RUnlock()
if item != nil {
  return item
}

// Probably not going to be in the cache, so create
// these while we aren't locking
expires := time.Now().Add(duration).UnixNano()
new_item := newItem(key, value, expires, track)

b.Lock()
defer b.Unlock()
// check again under write lock
item = b.lookup[key]
if item != nil {
  return item
}

b.lookup[key] = new_item
return item
```

Thoughts?

Reply:
done, i used the second method. 😁